### PR TITLE
Mix ephemeral-static DH output into MAC keys for v2

### PR DESCRIPTION
**Pull request by akalin-keybase**
Also mix in index into MAC nonce.

This prevents tampering of previously produced message ciphertexts,
even if the sender's private key is compromised.

### CODE REVIEW

File: common.go [hunk 1]
```diff
@@ -81,35 +84,94 @@ func detachedSignatureInputFromHash(plaintextAndHeaderHash []byte) []byte {
...
+	var auth payloadAuthenticator
+	h := authenticatorDigest.Sum(nil)
+	copyEqualSize(auth[:], h)
+	return auth
 }
 
-func computeMACKey(secret BoxSecretKey, public BoxPublicKey, headerHash []byte) []byte {
+type macKey [cryptoAuthKeyBytes]byte
+
+func computeMACKeySingle(secret BoxSecretKey, public BoxPublicKey, headerHash headerHash) macKey {
```
Comment (akalin-keybase) [inline common.go]: these `computeMacKey` functions comprise the meat of this PR.

**Comment by akalin-keybase** [comment]
This PR kinda sprawled a bit, so I might try to break it up. But interested in whether I'm on the right track, and if you have any strong opinions about any of the refactorings.

**Comment by oconnor663** [comment]
Thanks Fred! Will get to this this afternoon.

File: common.go [hunk 2]
```diff
@@ -81,35 +84,94 @@ func detachedSignatureInputFromHash(plaintextAndHeaderHash []byte) []byte {
...
+
+type payloadAuthenticator [sha512.Size256]byte
+
+func (pa payloadAuthenticator) Equal(other payloadAuthenticator) bool {
+	return hmac.Equal(pa[:], other[:])
+}
+
+func authenticatePayload(macKey macKey, payloadHash payloadHash) payloadAuthenticator {
+	// Equivalent to crypto_auth, but using Go's builtin HMAC.
+	authenticatorDigest := hmac.New(sha512.New512_256, macKey[:])
```
Comment (oconnor663) [inline common.go]: This is using a different hash function from before I think. Was that intentional?

File: const.go [hunk 3]
```diff
@@ -44,9 +44,9 @@ const SignedArmorString = "SIGNED MESSAGE"
 // DetachedSignatureArmorString is included in armor headers for detached signatures.
 const DetachedSignatureArmorString = "DETACHED SIGNATURE"
 
-// SaltpackFormatName is the publicly advertised name of the format,
-// used in the header of the message and also in Nonce creation.
-const SaltpackFormatName = "saltpack"
+// FormatName is the publicly advertised name of the format, used in
+// the header of the message and also in Nonce creation.
+const FormatName = "saltpack"
```
Comment (oconnor663) [inline const.go]: Possible pre-existing problem: I don't think we check the format name during decryption. Should we be doing that? I guess it depends on whether we expect someone to change the format name as a sort of context string (where accidental decryption compatibility could be a security issue) or as part of extending it in their own application (where it might be a jerk move to break them).

File: common.go [hunk 4]
```diff
@@ -81,35 +84,94 @@ func detachedSignatureInputFromHash(plaintextAndHeaderHash []byte) []byte {
...
+	h := authenticatorDigest.Sum(nil)
+	copyEqualSize(auth[:], h)
+	return auth
 }
 
-func computeMACKey(secret BoxSecretKey, public BoxPublicKey, headerHash []byte) []byte {
+type macKey [cryptoAuthKeyBytes]byte
+
+func computeMACKeySingle(secret BoxSecretKey, public BoxPublicKey, headerHash headerHash) macKey {
 	nonce := nonceForMACKeyBox(headerHash)
```
Comment (oconnor663) [inline common.go]: I think we're going to want to increment this nonce in V2, to avoid the problem where two recipients compute the same MAC key and then have visibly identical authenticators.

File: encrypt.go [hunk 5]
```diff
@@ -151,7 +156,8 @@ func (es *encryptStream) init(sender BoxSecretKey, receivers []BoxPublicKey) err
 	eh.SenderSecretbox = secretbox.Seal([]byte{}, sender.GetPublicKey().ToKID(), (*[24]byte)(nonceForSenderKeySecretBox()), (*[32]byte)(&es.payloadKey))
 
 	for _, receiver := range receivers {
-		payloadKeyBox := ephemeralKey.Box(receiver, nonceForPayloadKeyBoxV1(), es.payloadKey[:])
+		sharedKey := ephemeralKey.Precompute(receiver)
+		payloadKeyBox := sharedKey.Box(nonceForPayloadKeyBoxV1(), es.payloadKey[:])
```
Comment (oconnor663) [inline encrypt.go]: I think we're going to want to increment the nonce here too in V2, for the same reason.

File: common.go (hunk 2 above)
Comment (akalin-keybase) [inline common.go]: The old one used SHA512 and truncated it manually, right? Isn't this equivalent?

Comment (oconnor663) [inline common.go]: I don't think so. SHA-512/256 uses different initialization vectors. We can test it to see.

File: common.go (hunk 4 above)
Comment (akalin-keybase) [inline common.go]: Ah, yes, will do.

File: encrypt.go (hunk 5 above)
Comment (akalin-keybase) [inline encrypt.go]: This is the other issue that Trevor mentioned, right? I was planning to do that in another PR (plus possibly randomizing the order, etc.)

File: const.go (hunk 3 above)
Comment (akalin-keybase) [inline const.go]: I lean towards checking for the format name, and then later adding an option to remove/change the check if needed.

**Comment by akalin-keybase** [comment]
Okay, this is ready for review again!

File: common.go [hunk 6]
```diff
@@ -141,12 +141,58 @@ func computePayloadAuthenticator(macKey macKey, payloadHash payloadHash) payload
...
+
+func computeMACKeySender(version Version, index uint64, secret, eSecret BoxSecretKey, public BoxPublicKey, headerHash headerHash) macKey {
+	switch version {
+	case Version1():
+		nonce := nonceForMACKeyBoxV1(headerHash)
+		return computeMACKeySingle(secret, public, nonce)
+	case Version2():
+		nonce := nonceForMACKeyBoxV2(headerHash, index)
+		mac1 := computeMACKeySingle(secret, public, nonce)
+		mac2 := computeMACKeySingle(eSecret, public, nonce)
```
Comment (oconnor663) [inline common.go]: Would it be good practice to use a different nonce space for `mac2`? The `secret` and `eSecret` keys _are_ the same, when you're in anonymous sender mode, so this counts as nonce reuse. It does seem safe in this case, but then again...those are famous last words :-D

File: common.go [hunk 7]
```diff
@@ -141,12 +141,58 @@ func computePayloadAuthenticator(macKey macKey, payloadHash payloadHash) payload
...
+
+func computeMACKeyReceiver(version Version, index uint64, secret BoxSecretKey, public, ePublic BoxPublicKey, headerHash headerHash) macKey {
+	switch version {
+	case Version1():
+		nonce := nonceForMACKeyBoxV1(headerHash)
+		return computeMACKeySingle(secret, public, nonce)
+	case Version2():
+		nonce := nonceForMACKeyBoxV2(headerHash, index)
+		mac1 := computeMACKeySingle(secret, public, nonce)
+		mac2 := computeMACKeySingle(secret, ePublic, nonce)
```
Comment (oconnor663) [inline common.go]: Same here.

File: encrypt.go (hunk 5 above)
Comment (oconnor663) [inline encrypt.go]: The incrementing nonce is now in this PR, right?

Comment (oconnor663) [inline encrypt.go]: Or just for the MAC keys so far?

File: nonce.go [hunk 8]
```diff
@@ -31,11 +32,18 @@ func nonceForDerivedSharedKey() *Nonce {
...
 
-func nonceForMACKeyBox(headerHash headerHash) *Nonce {
+func nonceForMACKeyBoxV1(headerHash headerHash) *Nonce {
 	var n Nonce = sliceToByte24(headerHash[:nonceBytes])
 	return &n
 }
 
+func nonceForMACKeyBoxV2(headerHash headerHash, recip uint64) *Nonce {
+	var n Nonce
+	copyEqualSize(n[:nonceBytes-8], headerHash[:nonceBytes-8])
```
Comment (oconnor663) [inline nonce.go]: You could use `len(n)` instead of `nonceBytes` if you think that'll be clearer.

File: encrypt.go (hunk 5 above)
Comment (akalin-keybase) [inline encrypt.go]: Just for the MAC keys.

File: common.go (hunk 6 above)
Comment (akalin-keybase) [inline common.go]: Ooh, nice catch. Done!

File: common.go (hunk 7 above)
Comment (akalin-keybase) [inline common.go]: Done.

File: nonce.go (hunk 8 above)
Comment (akalin-keybase) [inline nonce.go]: Done.

**Comment by akalin-keybase** [comment]
PTAL!
